File: backend/app/models/collaboration.py
from sqlalchemy import Column, String, Integer, Boolean, DateTime, Text, JSON, ForeignKey, Enum, Float
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.postgresql import UUID
from datetime import datetime
import uuid
import enum

from ..core.database import Base

class ProjectRole(enum.Enum):
    LEADER = "leader"
    FRONTEND_DEVELOPER = "frontend_developer"
    BACKEND_DEVELOPER = "backend_developer"
    UI_UX_DESIGNER = "ui_ux_designer"
    DATA_SCIENTIST = "data_scientist"
    DEVOPS_ENGINEER = "devops_engineer"
    CONTRIBUTOR = "contributor"

class ParticipantStatus(enum.Enum):
    ONLINE = "online"
    AWAY = "away"
    BUSY = "busy"
    OFFLINE = "offline"

class ProjectPermission(enum.Enum):
    READ = "read"
    WRITE = "write"
    ADMIN = "admin"
    MANAGE_USERS = "manage_users"
    MANAGE_ROLES = "manage_roles"

class TeamProject(Base):
    __tablename__ = "team_projects"
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    name = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    created_by = Column(String, nullable=False, index=True)  # User ID of project leader
    
    # Project settings
    difficulty_level = Column(String, nullable=False, default="intermediate")  # beginner, intermediate, advanced
    max_team_size = Column(Integer, default=5)
    min_team_size = Column(Integer, default=2)
    is_public = Column(Boolean, default=True)
    requires_approval = Column(Boolean, default=False)
    
    # Project state
    status = Column(String, nullable=False, default="active")  # active, completed, paused, cancelled
    progress_percentage = Column(Integer, default=0)
    
    # Collaboration settings
    allow_voice_chat = Column(Boolean, default=True)
    allow_screen_share = Column(Boolean, default=True)
    auto_save_interval = Column(Integer, default=30)  # seconds
    
    # Metadata
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    completed_at = Column(DateTime, nullable=True)
    
    # Relationships
    participants = relationship("ProjectParticipant", back_populates="project", cascade="all, delete-orphan")
    sessions = relationship("CollaborationSession", back_populates="project", cascade="all, delete-orphan")
    files = relationship("ProjectFile", back_populates="project", cascade="all, delete-orphan")

class ProjectParticipant(Base):
    __tablename__ = "project_participants"
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    project_id = Column(String, ForeignKey("team_projects.id"), nullable=False)
    user_id = Column(String, nullable=False, index=True)
    
    # Role and permissions
    role = Column(Enum(ProjectRole), nullable=False, default=ProjectRole.CONTRIBUTOR)
    permissions = Column(JSON, default=list)  # List of ProjectPermission enums
    
    # Participation details
    joined_at = Column(DateTime, default=datetime.utcnow)
    last_active = Column(DateTime, default=datetime.utcnow)
    status = Column(Enum(ParticipantStatus), default=ParticipantStatus.OFFLINE)
    
    # Contribution tracking
    lines_contributed = Column(Integer, default=0)
    commits_made = Column(Integer, default=0)
    hours_contributed = Column(Float, default=0.0)
    
    # Relationships
    project = relationship("TeamProject", back_populates="participants")
    cursor_positions = relationship("CursorPosition", back_populates="participant", cascade="all, delete-orphan")

class CollaborationSession(Base):
    __tablename__ = "collaboration_sessions"
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    project_id = Column(String, ForeignKey("team_projects.id"), nullable=False)
    session_name = Column(String, nullable=True)  # Optional session name
    
    # Session state
    is_active = Column(Boolean, default=True)
    started_at = Column(DateTime, default=datetime.utcnow)
    ended_at = Column(DateTime, nullable=True)
    
    # Active participants
    active_participants = Column(JSON, default=list)  # List of user IDs currently in session
    active_files = Column(JSON, default=list)  # List of file IDs currently being edited
    
    # Communication state
    voice_channel_active = Column(Boolean, default=False)
    screen_share_active = Column(Boolean, default=False)
    screen_share_user = Column(String, nullable=True)  # User ID of person sharing screen
    
    # Relationships
    project = relationship("TeamProject", back_populates="sessions")
    messages = relationship("ChatMessage", back_populates="session", cascade="all, delete-orphan")

class ProjectFile(Base):
    __tablename__ = "project_files"
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    project_id = Column(String, ForeignKey("team_projects.id"), nullable=False)
    
    # File details
    filename = Column(String, nullable=False)
    file_path = Column(String, nullable=False)
    file_type = Column(String, nullable=False)  # file extension
    language = Column(String, nullable=True)  # programming language
    
    # Content and versioning
    content = Column(Text, nullable=True)
    version = Column(Integer, default=1)
    last_modified_by = Column(String, nullable=True)
    last_modified_at = Column(DateTime, default=datetime.utcnow)
    
    # File permissions
    is_locked = Column(Boolean, default=False)
    locked_by = Column(String, nullable=True)  # User ID of person who locked the file
    locked_at = Column(DateTime, nullable=True)
    
    # Relationships
    project = relationship("TeamProject", back_populates="files")
    changes = relationship("FileChange", back_populates="file", cascade="all, delete-orphan")

class FileChange(Base):
    __tablename__ = "file_changes"
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    file_id = Column(String, ForeignKey("project_files.id"), nullable=False)
    user_id = Column(String, nullable=False)
    
    # Change details
    change_type = Column(String, nullable=False)  # insert, delete, replace
    start_position = Column(Integer, nullable=False)
    end_position = Column(Integer, nullable=False)
    old_content = Column(Text, nullable=True)
    new_content = Column(Text, nullable=True)
    
    # Operational Transform data
    operation_id = Column(String, nullable=False)  # Unique ID for this operation
    parent_operation_id = Column(String, nullable=True)  # ID of parent operation for conflict resolution
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    applied_at = Column(DateTime, nullable=True)
    
    # Relationships
    file = relationship("ProjectFile", back_populates="changes")

class CursorPosition(Base):
    __tablename__ = "cursor_positions"
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    participant_id = Column(String, ForeignKey("project_participants.id"), nullable=False)
    file_id = Column(String, ForeignKey("project_files.id"), nullable=False)
    
    # Position data
    line = Column(Integer, nullable=False)
    column = Column(Integer, nullable=False)
    selection_start = Column(Integer, nullable=True)
    selection_end = Column(Integer, nullable=True)
    
    # Timestamps
    updated_at = Column(DateTime, default=datetime.utcnow)
    
    # Relationships
    participant = relationship("ProjectParticipant", back_populates="cursor_positions")

class ChatMessage(Base):
    __tablename__ = "chat_messages"
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    session_id = Column(String, ForeignKey("collaboration_sessions.id"), nullable=False)
    user_id = Column(String, nullable=False)
    
    # Message content
    message_type = Column(String, nullable=False, default="text")  # text, system, file_shared, etc.
    content = Column(Text, nullable=False)
    message_metadata = Column(JSON, default=dict)  # Additional message data
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    edited_at = Column(DateTime, nullable=True)
    
    # Relationships
    session = relationship("CollaborationSession", back_populates="messages")

class TeamInvitation(Base):
    __tablename__ = "team_invitations"
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    project_id = Column(String, ForeignKey("team_projects.id"), nullable=False)
    invited_user_id = Column(String, nullable=False, index=True)
    invited_by = Column(String, nullable=False)
    
    # Invitation details
    role = Column(Enum(ProjectRole), nullable=False, default=ProjectRole.CONTRIBUTOR)
    message = Column(Text, nullable=True)
    
    # Status
    status = Column(String, nullable=False, default="pending")  # pending, accepted, declined, expired
    expires_at = Column(DateTime, nullable=True)
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    responded_at = Column(DateTime, nullable=True)

class CollaborationStats(Base):
    __tablename__ = "collaboration_stats"
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = Column(String, nullable=False, index=True)
    project_id = Column(String, ForeignKey("team_projects.id"), nullable=False)
    
    # Session statistics
    total_sessions = Column(Integer, default=0)
    total_hours = Column(Float, default=0.0)
    last_session_at = Column(DateTime, nullable=True)
    
    # Contribution statistics
    lines_written = Column(Integer, default=0)
    lines_deleted = Column(Integer, default=0)
    files_created = Column(Integer, default=0)
    files_modified = Column(Integer, default=0)
    
    # Communication statistics
    messages_sent = Column(Integer, default=0)
    voice_minutes = Column(Float, default=0.0)
    screen_share_minutes = Column(Float, default=0.0)
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

File: frontend/src/types/collaboration.ts
export enum ProjectRole {
  LEADER = 'leader',
  FRONTEND_DEVELOPER = 'frontend_developer',
  BACKEND_DEVELOPER = 'backend_developer',
  UI_UX_DESIGNER = 'ui_ux_designer',
  DATA_SCIENTIST = 'data_scientist',
  DEVOPS_ENGINEER = 'devops_engineer',
  CONTRIBUTOR = 'contributor'
}

export enum ParticipantStatus {
  ONLINE = 'online',
  AWAY = 'away',
  BUSY = 'busy',
  OFFLINE = 'offline'
}

export enum ProjectPermission {
  READ = 'read',
  WRITE = 'write',
  ADMIN = 'admin',
  MANAGE_USERS = 'manage_users',
  MANAGE_ROLES = 'manage_roles'
}

export interface TeamProject {
  id: string;
  name: string;
  description: string;
  difficulty_level: 'beginner' | 'intermediate' | 'advanced';
  max_team_size: number;
  min_team_size: number;
  is_public: boolean;
  status: 'active' | 'completed' | 'paused' | 'cancelled';
  progress_percentage: number;
  created_by: string;
  created_at: string;
  updated_at?: string;
  completed_at?: string;
  participants?: ProjectParticipant[];
}

export interface ProjectParticipant {
  user_id: string;
  role: ProjectRole;
  permissions: ProjectPermission[];
  status: ParticipantStatus;
  joined_at: string;
  last_active: string;
  lines_contributed: number;
  commits_made: number;
  hours_contributed: number;
}

export interface CollaborationSession {
  id: string;
  project_id: string;
  session_name?: string;
  is_active: boolean;
  started_at: string;
  ended_at?: string;
  active_participants: string[];
  active_files: string[];
  voice_channel_active: boolean;
  screen_share_active: boolean;
  screen_share_user?: string;
}

export interface ProjectFile {
  id: string;
  filename: string;
  file_path: string;
  file_type: string;
  language?: string;
  content: string;
  version: number;
  last_modified_by?: string;
  last_modified_at: string;
  is_locked: boolean;
  locked_by?: string;
  locked_at?: string;
}

export interface FileChange {
  id: string;
  file_id: string;
  user_id: string;
  change_type: 'insert' | 'delete' | 'replace';
  start_position: number;
  end_position: number;
  old_content?: string;
  new_content?: string;
  operation_id: string;
  parent_operation_id?: string;
  created_at: string;
  applied_at?: string;
}

export interface CursorPosition {
  id: string;
  participant_id: string;
  file_id: string;
  line: number;
  column: number;
  selection_start?: number;
  selection_end?: number;
  updated_at: string;
}

export interface ChatMessage {
  id: string;
  user_id: string;
  message_type: 'text' | 'system' | 'file_shared' | 'voice_note';
  content: string;
  metadata?: Record<string, any>;
  created_at: string;
  edited_at?: string;
}

export interface TeamInvitation {
  id: string;
  project_id: string;
  invited_user_id: string;
  invited_by: string;
  role: ProjectRole;
  message?: string;
  status: 'pending' | 'accepted' | 'declined' | 'expired';
  created_at: string;
  expires_at?: string;
  responded_at?: string;
}

export interface CollaborationStats {
  user_id: string;
  project_id: string;
  total_sessions: number;
  total_hours: number;
  last_session_at?: string;
  lines_written: number;
  lines_deleted: number;
  files_created: number;
  files_modified: number;
  messages_sent: number;
  voice_minutes: number;
  screen_share_minutes: number;
}

// WebSocket Event Types
export interface WebSocketEvent {
  type: string;
  timestamp: string;
}

export interface UserJoinedEvent extends WebSocketEvent {
  type: 'user_joined';
  user_id: string;
  project_id: string;
}

export interface UserLeftEvent extends WebSocketEvent {
  type: 'user_left';
  user_id: string;
  project_id: string;
}

export interface CodeChangeEvent extends WebSocketEvent {
  type: 'code_change';
  user_id: string;
  project_id: string;
  file_id: string;
  change: {
    type: 'insert' | 'delete' | 'replace';
    start: number;
    end: number;
    text: string;
  };
}

export interface CursorMoveEvent extends WebSocketEvent {
  type: 'cursor_move';
  user_id: string;
  project_id: string;
  file_id: string;
  position: {
    line: number;
    column: number;
    selection_start?: number;
    selection_end?: number;
  };
}

export interface FileSwitchEvent extends WebSocketEvent {
  type: 'file_switch';
  user_id: string;
  project_id: string;
  file_id: string;
}

export interface ChatMessageEvent extends WebSocketEvent {
  type: 'chat_message';
  user_id: string;
  project_id: string;
  content: string;
  message_type: string;
}

export interface VoiceJoinedEvent extends WebSocketEvent {
  type: 'voice_joined';
  user_id: string;
  project_id: string;
}

export interface VoiceLeftEvent extends WebSocketEvent {
  type: 'voice_left';
  user_id: string;
  project_id: string;
}

export interface ScreenShareStartedEvent extends WebSocketEvent {
  type: 'screen_share_started';
  user_id: string;
  project_id: string;
}

export interface ScreenShareStoppedEvent extends WebSocketEvent {
  type: 'screen_share_stopped';
  user_id: string;
  project_id: string;
}

// Union type for all WebSocket events
export type CollaborationEvent = 
  | UserJoinedEvent
  | UserLeftEvent
  | CodeChangeEvent
  | CursorMoveEvent
  | FileSwitchEvent
  | ChatMessageEvent
  | VoiceJoinedEvent
  | VoiceLeftEvent
  | ScreenShareStartedEvent
  | ScreenShareStoppedEvent;

// Team Formation Types
export interface TeamMatchingCriteria {
  skill_level: 'beginner' | 'intermediate' | 'advanced';
  preferred_roles: ProjectRole[];
  timezone: string;
  availability_hours: number;
  project_difficulty: 'beginner' | 'intermediate' | 'advanced';
  team_size_preference: number;
}

export interface TeamMatch {
  user_id: string;
  compatibility_score: number;
  matching_skills: string[];
  timezone_compatibility: number;
  role_fit: ProjectRole;
}

// Project Creation Wizard Types
export interface ProjectCreationStep {
  step: number;
  title: string;
  description: string;
  fields: ProjectCreationField[];
}

export interface ProjectCreationField {
  name: string;
  type: 'text' | 'textarea' | 'select' | 'number' | 'checkbox';
  label: string;
  placeholder?: string;
  required: boolean;
  options?: Array<{ value: string; label: string }>;
  validation?: {
    min?: number;
    max?: number;
    pattern?: string;
  };
}

export interface ProjectCreationData {
  name: string;
  description: string;
  difficulty_level: 'beginner' | 'intermediate' | 'advanced';
  max_team_size: number;
  min_team_size: number;
  is_public: boolean;
  required_roles: ProjectRole[];
  skills_needed: string[];
  estimated_duration: number; // in days
  project_goals: string[];
}
